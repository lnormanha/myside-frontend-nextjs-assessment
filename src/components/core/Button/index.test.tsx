import "@testing-library/jest-dom";
import "jest-styled-components";
import { render } from "@testing-library/react";
import { Button } from "./index";
import { ServerStyleSheet } from "styled-components";

describe("Button", () => {
  let sheet: ServerStyleSheet;
  beforeEach(() => {
    sheet = new ServerStyleSheet();
    jest.spyOn(sheet, "collectStyles");
  });

  it("Should render a button", () => {
    const { container } = render(<Button>Click me</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe("Props - Size", () => {
    it("Should render a button with a small size", () => {
      const { container } = render(<Button size="small">Click me</Button>);
      expect(container.firstChild).toHaveStyleRule("height", "30px");
      expect(container.firstChild).toMatchSnapshot();
    });

    it("Should render a button with a medium size", () => {
      const { container } = render(<Button size="medium">Click me</Button>);
      expect(container.firstChild).toHaveStyleRule("height", "40px");
      expect(container.firstChild).toMatchSnapshot();
    });

    it("Should render a button with a large size", () => {
      const { container } = render(<Button size="large">Click me</Button>);
      expect(container.firstChild).toHaveStyleRule("height", "50px");
      expect(container.firstChild).toMatchSnapshot();
    });

    it("Should render a button with a full size", () => {
      const { container } = render(<Button size="full">Click me</Button>);
      expect(container.firstChild).toHaveStyleRule("width", "100%");
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe("Props - Variant", () => {
    it("Should render a button with a primary variant", () => {
      const { container } = render(<Button variant="primary">Click me</Button>);
      expect(container.firstChild).toHaveStyleRule("background-color", "#FFF");
      expect(container.firstChild).toMatchSnapshot();
    });

    it("Should render a button with a secondary variant", () => {
      const { container } = render(
        <Button variant="secondary">Click me</Button>
      );
      expect(container.firstChild).toHaveStyleRule(
        "background-color",
        "#303030"
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it("Should render a button with a outline variant", () => {
      const { container } = render(<Button variant="outline">Click me</Button>);
      expect(container.firstChild).toHaveStyleRule(
        "background-color",
        "transparent"
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe("Props - Disabled", () => {
    it("Should render a disabled button", () => {
      const { container } = render(<Button disabled>Click me</Button>);
      expect(container.firstChild).toHaveStyleRule("opacity", "0.5", {
        modifier: ":disabled",
      });
      expect(container.firstChild).toHaveStyleRule("cursor", "not-allowed", {
        modifier: ":disabled",
      });
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe("Props - Bg Color", () => {
    it("Should render a button with a bg color", () => {
      const { container } = render(<Button bgColor="red">Click me</Button>);
      expect(container.firstChild).toHaveStyleRule("background-color", "red");
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe("Props - Hover Bg Color", () => {
    it("Should render a button with a hover bg color", () => {
      const { container } = render(
        <Button hoverBgColor="red">Click me</Button>
      );
      expect(container.firstChild).toHaveStyleRule("background-color", "red", {
        modifier: ":hover",
      });
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
