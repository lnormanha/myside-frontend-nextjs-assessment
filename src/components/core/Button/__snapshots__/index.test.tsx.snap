// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Props - Bg Color Should render a button with a bg color 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: red;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  padding: 8px 16px;
  height: 40px;
  --text-size: 16px;
  --text-weight: 400;
  height: 40px;
  width: 200px;
  --text-size: 16px;
  --text-weight: 400;
  background-color: red;
  color: #000;
}

.c0:hover {
  background-color: #000;
  --text-color: "#fff";
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0:hover {
  background-color: #8a8a8a;
  --text-color: #525252;
}

.c1 {
  font-family: "Geist",sans-serif;
  color: var(--text-color,#000);
  text-align: left;
  font-weight: var(--text-weight,bold);
  font-size: var(--text-size,16px);
  --text-size: 16px;
}

<button
  bgcolor="red"
  class="c0"
  data-testid="button"
>
  <p
    class="c1"
    color="#000"
    data-testid="core-text-variant-p"
    weight="bold"
  >
    Click me
  </p>
</button>
`;

exports[`Button Props - Disabled Should render a disabled button 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #000;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  padding: 8px 16px;
  height: 40px;
  --text-size: 16px;
  --text-weight: 400;
  height: 40px;
  width: 200px;
  --text-size: 16px;
  --text-weight: 400;
  background-color: #FFF;
  color: #000;
}

.c0:hover {
  background-color: #000;
  --text-color: "#fff";
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0:hover {
  background-color: #8a8a8a;
  --text-color: #525252;
}

.c1 {
  font-family: "Geist",sans-serif;
  color: var(--text-color,#000);
  text-align: left;
  font-weight: var(--text-weight,bold);
  font-size: var(--text-size,16px);
  --text-size: 16px;
}

<button
  class="c0"
  data-testid="button"
  disabled=""
>
  <p
    class="c1"
    color="#000"
    data-testid="core-text-variant-p"
    weight="bold"
  >
    Click me
  </p>
</button>
`;

exports[`Button Props - Hover Bg Color Should render a button with a hover bg color 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #000;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  padding: 8px 16px;
  height: 40px;
  --text-size: 16px;
  --text-weight: 400;
  height: 40px;
  width: 200px;
  --text-size: 16px;
  --text-weight: 400;
  background-color: #FFF;
  color: #000;
}

.c0:hover {
  background-color: red;
  --text-color: "#fff";
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0:hover {
  background-color: red;
  --text-color: #525252;
}

.c1 {
  font-family: "Geist",sans-serif;
  color: var(--text-color,#000);
  text-align: left;
  font-weight: var(--text-weight,bold);
  font-size: var(--text-size,16px);
  --text-size: 16px;
}

<button
  class="c0"
  data-testid="button"
  hoverbgcolor="red"
>
  <p
    class="c1"
    color="#000"
    data-testid="core-text-variant-p"
    weight="bold"
  >
    Click me
  </p>
</button>
`;

exports[`Button Props - Size Should render a button with a full size 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #000;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  padding: 8px 16px;
  height: 40px;
  --text-size: 16px;
  --text-weight: 400;
  width: 100%;
  background-color: #FFF;
  color: #000;
}

.c0:hover {
  background-color: #000;
  --text-color: "#fff";
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0:hover {
  background-color: #8a8a8a;
  --text-color: #525252;
}

.c1 {
  font-family: "Geist",sans-serif;
  color: var(--text-color,#000);
  text-align: left;
  font-weight: var(--text-weight,bold);
  font-size: var(--text-size,16px);
  --text-size: 16px;
}

<button
  class="c0"
  data-testid="button"
>
  <p
    class="c1"
    color="#000"
    data-testid="core-text-variant-p"
    weight="bold"
  >
    Click me
  </p>
</button>
`;

exports[`Button Props - Size Should render a button with a large size 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #000;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  padding: 8px 16px;
  height: 40px;
  --text-size: 16px;
  --text-weight: 400;
  height: 50px;
  width: 300px;
  --text-size: 20px;
  --text-weight: 500;
  background-color: #FFF;
  color: #000;
}

.c0:hover {
  background-color: #000;
  --text-color: "#fff";
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0:hover {
  background-color: #8a8a8a;
  --text-color: #525252;
}

.c1 {
  font-family: "Geist",sans-serif;
  color: var(--text-color,#000);
  text-align: left;
  font-weight: var(--text-weight,bold);
  font-size: var(--text-size,16px);
  --text-size: 16px;
}

<button
  class="c0"
  data-testid="button"
>
  <p
    class="c1"
    color="#000"
    data-testid="core-text-variant-p"
    weight="bold"
  >
    Click me
  </p>
</button>
`;

exports[`Button Props - Size Should render a button with a medium size 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #000;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  padding: 8px 16px;
  height: 40px;
  --text-size: 16px;
  --text-weight: 400;
  height: 40px;
  width: 200px;
  --text-size: 16px;
  --text-weight: 400;
  background-color: #FFF;
  color: #000;
}

.c0:hover {
  background-color: #000;
  --text-color: "#fff";
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0:hover {
  background-color: #8a8a8a;
  --text-color: #525252;
}

.c1 {
  font-family: "Geist",sans-serif;
  color: var(--text-color,#000);
  text-align: left;
  font-weight: var(--text-weight,bold);
  font-size: var(--text-size,16px);
  --text-size: 16px;
}

<button
  class="c0"
  data-testid="button"
>
  <p
    class="c1"
    color="#000"
    data-testid="core-text-variant-p"
    weight="bold"
  >
    Click me
  </p>
</button>
`;

exports[`Button Props - Size Should render a button with a small size 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #000;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  padding: 8px 16px;
  height: 40px;
  --text-size: 16px;
  --text-weight: 400;
  height: 30px;
  width: 100px;
  --text-size: 12px;
  --text-weight: 400;
  background-color: #FFF;
  color: #000;
}

.c0:hover {
  background-color: #000;
  --text-color: "#fff";
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0:hover {
  background-color: #8a8a8a;
  --text-color: #525252;
}

.c1 {
  font-family: "Geist",sans-serif;
  color: var(--text-color,#000);
  text-align: left;
  font-weight: var(--text-weight,bold);
  font-size: var(--text-size,16px);
  --text-size: 16px;
}

<button
  class="c0"
  data-testid="button"
>
  <p
    class="c1"
    color="#000"
    data-testid="core-text-variant-p"
    weight="bold"
  >
    Click me
  </p>
</button>
`;

exports[`Button Props - Variant Should render a button with a outline variant 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #000;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  padding: 8px 16px;
  height: 40px;
  --text-size: 16px;
  --text-weight: 400;
  height: 40px;
  width: 200px;
  --text-size: 16px;
  --text-weight: 400;
  background-color: transparent;
  color: #ffffff;
  border: 1px solid #ffffff;
}

.c0:hover {
  background-color: #000;
  --text-color: "#fff";
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0:hover {
  background-color: #ffffff0c;
  --text-color: #fff;
  border: 1px solid #ffffff;
}

.c1 {
  font-family: "Geist",sans-serif;
  color: var(--text-color,#FFF);
  text-align: left;
  font-weight: var(--text-weight,bold);
  font-size: var(--text-size,16px);
  --text-size: 16px;
}

<button
  class="c0"
  data-testid="button"
>
  <p
    class="c1"
    color="#FFF"
    data-testid="core-text-variant-p"
    weight="bold"
  >
    Click me
  </p>
</button>
`;

exports[`Button Props - Variant Should render a button with a primary variant 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #000;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  padding: 8px 16px;
  height: 40px;
  --text-size: 16px;
  --text-weight: 400;
  height: 40px;
  width: 200px;
  --text-size: 16px;
  --text-weight: 400;
  background-color: #FFF;
  color: #000;
}

.c0:hover {
  background-color: #000;
  --text-color: "#fff";
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0:hover {
  background-color: #8a8a8a;
  --text-color: #525252;
}

.c1 {
  font-family: "Geist",sans-serif;
  color: var(--text-color,#000);
  text-align: left;
  font-weight: var(--text-weight,bold);
  font-size: var(--text-size,16px);
  --text-size: 16px;
}

<button
  class="c0"
  data-testid="button"
>
  <p
    class="c1"
    color="#000"
    data-testid="core-text-variant-p"
    weight="bold"
  >
    Click me
  </p>
</button>
`;

exports[`Button Props - Variant Should render a button with a secondary variant 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #000;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  padding: 8px 16px;
  height: 40px;
  --text-size: 16px;
  --text-weight: 400;
  height: 40px;
  width: 200px;
  --text-size: 16px;
  --text-weight: 400;
  background-color: #303030;
  color: #FFF;
}

.c0:hover {
  background-color: #000;
  --text-color: "#fff";
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0:hover {
  background-color: #1a1a1a;
  --text-color: #fff;
}

.c1 {
  font-family: "Geist",sans-serif;
  color: var(--text-color,#FFF);
  text-align: left;
  font-weight: var(--text-weight,bold);
  font-size: var(--text-size,16px);
  --text-size: 16px;
}

<button
  class="c0"
  data-testid="button"
>
  <p
    class="c1"
    color="#FFF"
    data-testid="core-text-variant-p"
    weight="bold"
  >
    Click me
  </p>
</button>
`;

exports[`Button Should render a button 1`] = `
.c0 {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #000;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  padding: 8px 16px;
  height: 40px;
  --text-size: 16px;
  --text-weight: 400;
  height: 40px;
  width: 200px;
  --text-size: 16px;
  --text-weight: 400;
  background-color: #FFF;
  color: #000;
}

.c0:hover {
  background-color: #000;
  --text-color: "#fff";
}

.c0:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0:hover {
  background-color: #8a8a8a;
  --text-color: #525252;
}

.c1 {
  font-family: "Geist",sans-serif;
  color: var(--text-color,#000);
  text-align: left;
  font-weight: var(--text-weight,bold);
  font-size: var(--text-size,16px);
  --text-size: 16px;
}

<button
  class="c0"
  data-testid="button"
>
  <p
    class="c1"
    color="#000"
    data-testid="core-text-variant-p"
    weight="bold"
  >
    Click me
  </p>
</button>
`;
